service/storage_service.cc

```diff
@@ -6038,6 +6077,187 @@ future<> storage_service::stream_tablet(locator::global_tablet_id tablet) {
+            if (it == _topology_state_machine._topology.normal_nodes.end()) {
+                result.result = join_node_request_result::rejected {
+                    .reason = ::format("Cannot replace node {} because it is not in the 'normal' state", *params.replaced_id),
+                };
+                co_return result;
+            }
+        }
+
+        // The node was not yet removed from the cluster
+
+        if (auto *p = _topology_state_machine._topology.find(params.host_id)) {
+            auto& rs = p->second;
+            if (!rs.join_cookie) {
+                // Return ok, for idempotency
+                result.result = join_node_request_result::ok {};
+                slogger.info("raft topology: join: returning ok for idempotency ({})", params.host_id);
+                co_return result;
+            } else {
+                assert(rs.state == node_state::none);
+                if (params.join_cookie != *rs.join_cookie) {
+                    // The topology coordinator still has to process the old request
+                    // and the node should reject it, only then the new request
+                    // can be placed
+                    result.result = join_node_request_result::in_progress {};
+                    slogger.info("raft topology: join: request to join {} is already placed", params.host_id);
+                    co_return result;
+                } else {
+                    // Return ok, for idempotency
+                    result.result = join_node_request_result::ok {};
+                    slogger.info("raft topology: join: returning ok for idempotency, ({})", params.host_id);
+                    co_return result;
+                }
+            }
+        }
+
+        // The node was not yet added to the cluster
+        // We can place the request now
+        auto mutation = build_mutation_from_join_params(params, guard);
+
+        topology_change change{{std::move(mutation)}};
+        group0_command g0_cmd = _group0->client().prepare_command(std::move(change), guard,
+                format("raft topology: adding join request for {}", params.host_id));
+        try {
+            co_await _group0->client().add_entry(std::move(g0_cmd), std::move(guard), &_abort_source);
+            break;
+        } catch (group0_concurrent_modification&) {
+            slogger.info("raft topology: join_node_request: concurrent operation is detected, retrying.");
+        }
+    }
+
+    slogger.info("raft topology: placed join request for {}", params.host_id);
+
+    // Success
+    result.result = join_node_request_result::ok {};
+    co_return result;
+}
+
+future<join_node_response_result> storage_service::join_node_response_handler(join_node_response_params params) {
+    assert(this_shard_id() == 0);
+
+    // Serialize handling the responses, just to be sure
+    auto lock = co_await get_units(_join_node_response_handler_mutex, 1);
```

Comment:
IIUC it could happen that we handle this message before we finish the two-way REQUEST RPC; is that OK?

Reply:
Should not happen now, the join_node_response handler now waits until join_node_request was sent and finished successfully.